### PR TITLE
Documentation: add a document on storage options

### PR DESCRIPTION
**Pull request by ericchiang**
fyi @dghubble there's some third party resource docs here.

### CODE REVIEW

File: Documentation/storage.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
+# Storage options
+
+Dex requires persisting state to perform various tasks such as track refresh tokens, preventing replays, and rotating keys. This document is a summary of the storage configurations supported by dex.
+
+Regardless of the implementation, dex saves sensitive data in its backing storage: signing keys, bcrypt'd passwords, etc. Storage breaches are serious because attacks may propagate to any application which relies on dex. Whatever storage option used, transport security and database ACLs are both musts.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword this to be:

**Note:** Storage breaches are serious as they can affect applications that rely on dex. Dex saves a lot of sensitive data in its backing storage, including **signing keys** and **bcrypt'd passwords**. As such, transport security and database ACLs should both be used, no matter which storage option is chosen. (LINK TO SETUP DOCS FOR THIS)

File: Documentation/storage.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+## Caveat: Running replicated instances
+
+Validation is still being done to ensure correctness when multiple instances of dex are using the same backing storage. While everything "should just work", edge cases require time to investigate and write tests for. This work can be followed on the issue tracker [here](https://github.com/coreos/dex/issues/600).
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword this to be:

Dex performs correctness validation when multiple instances use the same backing storage.  This should "Just Work" but when it doesn't you should check out the `[dex issue tracker][dex-issue-tracker-url]` and add or comment on a relevant issue.

File: Documentation/storage.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+For now, instances should run a single instance of dex and occationally run two during upgrades.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword this to be (if it's accurate):

The Dex team suggests running **one dex instance most of the time** and **two instance during upgrades**.

File: Documentation/storage.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+## Kubernetes third party resources
+
+NOTE: Dex requires Kubernetes version 1.4+.
+
+Kubernetes third party resources are a way for applications to create new resources types in the Kubernetes API. This allows dex to run on top of an existing Kubernetes cluster without the need for an external database. While this storage may not be appropriate for a large number of users, it's extremely effective for may Kubernetes use cases.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: [...] it's extremely effective for **many** Kubernetes use cases.

File: Documentation/storage.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    inCluster: true
+```
+
+Dex determines the namespace it's running in by parsing the service account token automatically mounted into its pod.
+
+## SQL
+
+Dex supports two flavors of SQL, SQLite3 and Postgres. MySQL and CockroachDB may be added at a later time.
+
+Migrations are performed automatically on the first connection to the SQL server (it does not support rolling back). Because of this, dex requires being able to add and alter the tables for its database.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword the second sentence to be:

Because of this dex requires privileges to add and alter tables for its database.

File: Documentation/storage.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+## Caveat: Running replicated instances
+
+Validation is still being done to ensure correctness when multiple instances of dex are using the same backing storage. While everything "should just work", edge cases require time to investigate and write tests for. This work can be followed on the issue tracker [here](https://github.com/coreos/dex/issues/600).
+
+For now, instances should run a single instance of dex and occationally run two during upgrades.
+
+## Kubernetes third party resources
+
+NOTE: Dex requires Kubernetes version 1.4+.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would make `NOTE` **bold**.

File: Documentation/storage.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+Dex determines the namespace it's running in by parsing the service account token automatically mounted into its pod.
+
+## SQL
+
+Dex supports two flavors of SQL, SQLite3 and Postgres. MySQL and CockroachDB may be added at a later time.
+
+Migrations are performed automatically on the first connection to the SQL server (it does not support rolling back). Because of this, dex requires being able to add and alter the tables for its database.
+
+NOTE: Previous versions of dex required symmetric keys to encrypt certain values before sending them to the database. This feature has not been ported to V2 yet, though it may be added later (but possibly without a migration path for users of the current implementation).
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would make `Note` **bold**.

Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword the last second sentence to be

This feature has not yet been ported to dex v2.  If it is added later there may not be a migration path for current v2 users.

File: Documentation/storage.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+  type: sqlite3
+  config:
+    file: /var/dex/dex.db
+```
+
+Because SQLite3 uses file locks to prevent race conditions, if the ":memory:" value is provided dex will automatically disable support for concurrent database queries.
+
+### Postgres
+
+Dex authenticates to Postgres with a username and password. As stated above, because dex performs migrations, it requires privileged access to its own database. Database table names are not configurable, and to prevent table name clashes with other applications admins should expect to dedicated a database for dex's sole use.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword this paragraph to be:

Admins may want to dedicate a database to dex for the following reasons:

1. Dex requires privileged access to its database because it performs migrations.
2. Dex's database table names are not configurable; when shared with other applications there may be table name clashes.

File: Documentation/storage.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+The SSL "mode" corresponds to the `github.com/lib/pq` package [connection options](https://godoc.org/github.com/lib/pq#hdr-Connection_String_Parameters). If unspecified, dex defaults to the strictest mode "verify-full".
+
+## Adding a new storage options
+
+Each storage implementation bears a large ongoing maintenance cost and needs to be updated every time a feature requires storing a new type. Bugs often require in depth knowledge of the backing software, and much of this work will be done by developers who are not the original authors. Changes to dex which add new storage implementations are not merged lightly.
+
+Those who still want to construct a proposal for a new storage should review the following packages:
+
+* `github.com/coreos/dex/storage`: Interface definitions which the storage must implement. NOTE: This package is not stable.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: This should be a clickable link.

File: Documentation/storage.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+* `github.com/coreos/dex/storage/conformance`: Conformance tests which storage implementations must pass.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: This should be a clickable link.

File: Documentation/storage.md [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+Any proposal to add a new implementation must address the following:
+
+* Integration testing setups (Travis and developer workstations).
+* Transactional requirements: atomic deletes, updates, etc.
+* Is there an established, reasonable Go client?
```
Comment (ElijahCaine) [inline Documentation/storage.md]: I would reword this last bullet to be:

* Is there an established and reasonable Go client?

File: Documentation/storage.md (hunk 1 above)
Comment (ericchiang) [inline Documentation/storage.md]: Thanks! Though this is the document that should have those setups.

File: Documentation/storage.md (hunk 2 above)
Comment (ericchiang) [inline Documentation/storage.md]: > Dex performs correctness validation when multiple instances use the same backing storage. 

That's not the same thing. This is stating that we still need to write the tests to prove correctness and you probably shouldn't be running these kind of setups in the first place. How does this sound?

> Tests still need to be written to validate that multiple instances of dex behave correctly when using the same storage.
> While there aren't any technical limitations, edge cases have been observed and progress on these kind of bugs can be found on the `[dex issue tracker]`.

File: Documentation/storage.md [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      caFile: /etc/dex/postgres.ca
+```
+
+The SSL "mode" corresponds to the `github.com/lib/pq` package [connection options](https://godoc.org/github.com/lib/pq#hdr-Connection_String_Parameters). If unspecified, dex defaults to the strictest mode "verify-full".
+
+## Adding a new storage options
+
+Each storage implementation bears a large ongoing maintenance cost and needs to be updated every time a feature requires storing a new type. Bugs often require in depth knowledge of the backing software, and much of this work will be done by developers who are not the original authors. Changes to dex which add new storage implementations are not merged lightly.
+
+Those who still want to construct a proposal for a new storage should review the following packages:
```
Comment (ericchiang) [inline Documentation/storage.md]: Wouldn't that imply that it's part of the SQL section (as it's not)?

File: Documentation/storage.md (hunk 9 above)
Comment (ericchiang) [inline Documentation/storage.md]: We can't link to a godoc.org until we've made this branch the default branch. Part of that getting to that point is writing docs like this :)

**Comment by ericchiang** [comment]
@ElijahCaine thanks. Have a few counter comments, but other than that will rework this PR later today.

File: Documentation/storage.md (hunk 9 above)
Comment (ElijahCaine) [inline Documentation/storage.md]: Fair enough.  I wanted to click on it and figured readers would too, but if it's technically not an option that's 👍

File: Documentation/storage.md (hunk 1 above)
Comment (ElijahCaine) [inline Documentation/storage.md]: Ah.  I think I read the thing about signing keys and didn't read anything about how to set that up (because it hasn't been ported yet). That's not required if this is the only doc that covers those topics.

**Comment by ericchiang** [comment]
@ElijahCaine updated with a second commit. Going to leave this here if you want to add any last second notes. Will squash and merge in a bit.

File: Documentation/storage.md [hunk 13]
```diff
@@ -121,7 +140,10 @@ Because SQLite3 uses file locks to prevent race conditions, if the ":memory:" va
 
 ### Postgres
 
-Dex authenticates to Postgres with a username and password. As stated above, because dex performs migrations, it requires privileged access to its own database. Database table names are not configurable, and to prevent table name clashes with other applications admins should expect to dedicated a database for dex's sole use. 
+When using Postgres, admins may want to dedicate a database to dex for the following reasons:
+
+1. Dex requires privileged access ot its database because it performs migrations.
+1. Dex's database table names are not configurable; when shared with other applications there may be table name clashes.
```
Comment (ElijahCaine) [inline Documentation/storage.md]: This should probably be `2.`

**Comment by ericchiang** [comment]
Haha, thanks I'll update again.

File: Documentation/storage.md (hunk 13 above)
Comment (ericchiang) [inline Documentation/storage.md]: updated, thanks
